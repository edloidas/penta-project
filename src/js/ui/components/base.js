// @flow

import styled from 'styled-components';

type Colors = { [string]: string };

// eslint-disable-next-line import/prefer-default-export
export const Fullscreen = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  margin: 0;
  box-sizing: border-box;
  overflow: hidden;
`;

// prettier-ignore
export const colors: Colors = {
  bg:           '#111',
  bgReverse:    '#f5f5f5',
  bgMenu:       '#1a1a1a',
  bgActive1:    '#1d1f21',
  bgActive2:    '#282a2e',
  bgActive3:    '#373b41',
  fontInactive: 'rgba(255, 255, 255, 0.2)',
  fontDesc:     '#c5c8c6',
  font:         '#fff',    // white
  accent:       '#b8860b', // darkgoldenrod
  hp:           '#b82739',
  eg:           '#08637e',
};
