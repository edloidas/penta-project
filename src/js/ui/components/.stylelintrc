{
  "processors": ["stylelint-processor-styled-components"],
  "extends": "stylelint-config-standard",
  "syntax": "scss",
  "rules": {
    "indentation": 2,
    "unit-whitelist": [
      [ "em", "rem", "%", "s" ],
      {
        "ignoreProperties": {
          "px": [ "background-position" ]
        }
      }
    ]
  }
}
