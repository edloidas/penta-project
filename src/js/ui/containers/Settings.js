// @flow
import React from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import clone from 'lodash.clonedeep';
import * as SettingsActions from '../actions/settings';
import { type SettingsData } from '../reducers/settings';
import SettingsMenu from '../components/SettingsMenu/SettingsMenu';
import { type NavigationItemProps } from '../components/Navigation/NavigationItem';
import { texts } from '../components/base';
import { type State } from '../store';
import Focusable from './Focusable';

function checkActivateEvent(event?: MouseKeyboardEvent) {
  if (event != null) {
    event.stopPropagation();
    event.preventDefault();
    const e = event.nativeEvent || event;
    const isMainButtonClicked = e instanceof MouseEvent && e.button === 0;
    const isMainKeyPressed =
      e instanceof KeyboardEvent && (e.code === 'Enter' || e.code === 'Space');
    return isMainButtonClicked || isMainKeyPressed;
  }
  return false;
}

type Props = {
  activeGroup: SettingsGroup,
  hasUnsavedChanges: boolean,
  data: SettingsData,
  actions: {
    switchSettings: typeof SettingsActions.switchSettings,
    setSettings: typeof SettingsActions.setSettings,
    applySettings: typeof SettingsActions.applySettings,
    resetSettings: typeof SettingsActions.resetSettings
  }
};

// eslint-disable-next-line react/prefer-stateless-function
class Settings extends Focusable<Props> {
  constructor(props) {
    super(props);
    this.handleSwitchGroup = this.handleSwitchGroup.bind(this);
    this.handleSetSettings = this.handleSetSettings.bind(this);
    this.handleApplySettings = this.handleApplySettings.bind(this);
    this.handleResetSettings = this.handleResetSettings.bind(this);

    const { settingsGroup } = texts;
    this.navigationItems = settingsGroup.map(name => ({
      name,
      dataKey: name,
      clickHandler: this.handleSwitchGroup
    }));
  }

  props: Props;

  handleSwitchGroup: (e?: MouseKeyboardEvent) => void;
  handleSetSettings: (value: string) => void;
  handleApplySettings: (e?: MouseKeyboardEvent) => void;
  handleResetSettings: (e?: MouseKeyboardEvent) => void;

  navigationItems: Array<NavigationItemProps>;

  handleSwitchGroup(e?: MouseKeyboardEvent) {
    if (checkActivateEvent(e) && e && e.currentTarget) {
      const key = e.currentTarget.getAttribute('data-key');
      const { settingsGroup } = texts;
      const group = settingsGroup.find(v => v === key) || settingsGroup[0];
      this.props.actions.switchSettings(group);
      e.currentTarget.blur();
    }
  }

  handleSetSettings(data: SettingsData) {
    this.props.actions.setSettings(data);
  }

  handleApplySettings(e?: MouseKeyboardEvent) {
    if (checkActivateEvent(e)) {
      this.props.actions.applySettings();
    }
  }

  handleResetSettings(e?: MouseKeyboardEvent) {
    if (checkActivateEvent(e)) {
      this.props.actions.resetSettings(this.props.data);
    }
  }

  render() {
    return (
      <div
        tabIndex="0"
        role="button"
        ref={div => {
          this.focusable = div;
        }}>
        <SettingsMenu
          handleApplySettings={this.handleApplySettings}
          handleResetSettings={this.handleResetSettings}
          navigationItems={this.navigationItems}
          activeKey={this.props.activeGroup}
        />
      </div>
    );
  }
}

function mapStateToProps(state: State) {
  const { settings } = state;
  return {
    activeGroup: settings.activeGroup,
    hasUnsavedChanges: settings.hasUnsavedChanges,
    data: clone(settings.data)
  };
}

function mapDispatchToProps(dispatch: Function) {
  return {
    actions: bindActionCreators(SettingsActions, dispatch)
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Settings);
