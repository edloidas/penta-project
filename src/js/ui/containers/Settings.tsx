import { cloneDeep } from 'lodash';
import * as React from 'react';
import { connect } from 'react-redux';
import { Dispatch, bindActionCreators } from 'redux';
import * as SettingsActions from '../actions/settings';
import { NavigationItemProps } from '../components/Navigation/NavigationItem';
import SettingsMenu from '../components/SettingsMenu/SettingsMenu';
import { texts } from '../components/base';
import { SettingsData } from '../reducers/settings';
import { State } from '../store';
import Focusable from './Focusable';

function checkActivateEvent(event?: MouseKeyboardEvent) {
  if (event != null) {
    event.stopPropagation();
    event.preventDefault();
    const e = event.nativeEvent || event;
    const isMainButtonClicked = e instanceof MouseEvent && e.button === 0;
    const isMainKeyPressed =
      e instanceof KeyboardEvent && (e.code === 'Enter' || e.code === 'Space');
    return isMainButtonClicked || isMainKeyPressed;
  }
  return false;
}

type Props = {
  activeGroup: SettingsGroup;
  hasUnsavedChanges: boolean;
  data: SettingsData;
  actions: {
    switchSettings: typeof SettingsActions.switchSettings;
    setSettings: typeof SettingsActions.setSettings;
    applySettings: typeof SettingsActions.applySettings;
    resetSettings: typeof SettingsActions.resetSettings;
  };
};

class Settings extends Focusable<Props> {
  props: Props;

  navigationItems: NavigationItemProps[];

  constructor(props) {
    super(props);
    this.handleSwitchGroup = this.handleSwitchGroup.bind(this);
    this.handleSetSettings = this.handleSetSettings.bind(this);
    this.handleReturnToMenu = this.handleReturnToMenu.bind(this);
    this.handleApplySettings = this.handleApplySettings.bind(this);
    this.handleResetSettings = this.handleResetSettings.bind(this);

    const { settingsGroup } = texts;
    this.navigationItems = settingsGroup.map(name => ({
      name,
      dataKey: name,
      clickHandler: this.handleSwitchGroup,
    }));
  }

  handleSwitchGroup(e?: MouseKeyboardEvent) {
    if (checkActivateEvent(e) && e && e.currentTarget) {
      const key = e.currentTarget.getAttribute('data-key');
      const { settingsGroup } = texts;
      const group = settingsGroup.find(v => v === key) || settingsGroup[0];
      this.props.actions.switchSettings(group);
      e.currentTarget.blur();
    }
  }

  handleSetSettings(data: SettingsData) {
    this.props.actions.setSettings(data);
  }

  handleReturnToMenu() {
    const { settingsGroup } = texts;
    this.props.actions.switchSettings(settingsGroup[0]);
  }

  handleApplySettings(e?: MouseKeyboardEvent) {
    if (checkActivateEvent(e)) {
      this.props.actions.applySettings();
      this.handleReturnToMenu();
    }
  }

  handleResetSettings(e?: MouseKeyboardEvent) {
    if (checkActivateEvent(e)) {
      this.props.actions.resetSettings(this.props.data);
      this.handleReturnToMenu();
    }
  }

  render() {
    return (
      <div>
        <SettingsMenu
          handleReturnToMenu={this.handleReturnToMenu}
          handleApplySettings={this.handleApplySettings}
          handleResetSettings={this.handleResetSettings}
          navigationItems={this.navigationItems}
          activeKey={this.props.activeGroup}
          hasUnsavedChanges={this.props.hasUnsavedChanges}
        />
      </div>
    );
  }
}

function mapStateToProps(state: State) {
  const { settings } = state;
  return {
    activeGroup: settings.activeGroup,
    hasUnsavedChanges: settings.hasUnsavedChanges,
    data: cloneDeep(settings.data),
  };
}

function mapDispatchToProps(dispatch: Dispatch<SettingsActions.ActionsTypes>) {
  return {
    actions: bindActionCreators(SettingsActions, dispatch),
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Settings);
