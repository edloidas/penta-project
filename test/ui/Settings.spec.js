import React from 'react';
import configureStore from '../../src/js/ui/store';
import Settings from '../../src/js/ui/containers/Settings';
import {
  switchSettings,
  setSettings,
  applySettings,
  resetSettings
} from '../../src/js/ui/actions/settings';
import { texts } from '../../src/js/ui/components/base';
import renderTree from './renderTree';

const initialData = {
  graphics: {
    screenSize: texts.graphics.screenSize[0]
  }
};

const data = {
  graphics: {
    screenSize: texts.graphics.screenSize[1]
  }
};

describe('<Settings />', () => {
  test('Should renders', () => {
    const store = configureStore();

    const tree = renderTree(store, <Settings />);
    expect(tree).toMatchSnapshot();
  });

  test('Should switch group', () => {
    const store = configureStore();
    store.dispatch(switchSettings(texts.settingsGroup[1]));

    const tree = renderTree(store, <Settings />);
    expect(tree).toMatchSnapshot();
  });

  test('Should update settings and switch to "changed" state', () => {
    const store = configureStore();
    store.dispatch(setSettings(data));

    const tree = renderTree(store, <Settings />);
    expect(tree).toMatchSnapshot();
  });

  test('Should reset settings', () => {
    const store = configureStore();

    const tree = renderTree(store, <Settings />);

    store.dispatch(setSettings(data));
    const treeWithChanges = renderTree(store, <Settings />);
    expect(treeWithChanges).toMatchSnapshot();

    store.dispatch(resetSettings(initialData));
    const treeWithoutChanges = renderTree(store, <Settings />);
    expect(treeWithoutChanges).toMatchSnapshot();

    expect(JSON.stringify(tree)).toEqual(JSON.stringify(treeWithoutChanges));
  });

  test('Should apply settings', () => {
    const store = configureStore();

    store.dispatch(setSettings(data));
    const treeWithChanges = renderTree(store, <Settings />);
    expect(treeWithChanges).toMatchSnapshot();

    store.dispatch(applySettings());
    const tree = renderTree(store, <Settings />);
    expect(tree).toMatchSnapshot();
  });
});
