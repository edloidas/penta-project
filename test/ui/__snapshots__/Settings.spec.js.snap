// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Settings /> Should apply settings 1`] = `
<div
  role="button"
  tabIndex="0"
>
  <div
    className="effect__appear sc-bwzfXH iVThND"
  >
    <div
      className="sc-bZQynM hKpJBw"
    >
      <h1
        className="font__accent sc-htpNat ggNPGu"
      >
        Settings
      </h1>
      <div
        className="font__accent sc-EHOje hPfYAa"
      >
        <button
          className="sc-ifAKCX hgRAff"
          data-key="graphics"
          onClick={[Function]}
          tabIndex={0}
        >
          graphics
        </button>
        <button
          className="sc-ifAKCX gXidXF"
          data-key="sound"
          onClick={[Function]}
          tabIndex={0}
        >
          sound
        </button>
        <button
          className="sc-ifAKCX gXidXF"
          data-key="controls"
          onClick={[Function]}
          tabIndex={0}
        >
          controls
        </button>
      </div>
      <div
        className="sc-gzVnrw jyBrwk"
      >
        <button
          className="effect__glitch sc-bxivhb eGYeuD"
          data-text="Apply"
          onClick={[Function]}
          tabIndex={-1}
        >
          Apply
          <a
            href="/menu"
            onClick={[Function]}
            tabIndex="0"
          />
        </button>
        <button
          className="effect__glitch sc-bxivhb eGYeuD"
          data-text="Reset"
          onClick={[Function]}
          tabIndex={-1}
        >
          Reset
          <a
            href="/menu"
            onClick={[Function]}
            tabIndex="0"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Settings /> Should apply settings 2`] = `
<div
  role="button"
  tabIndex="0"
>
  <div
    className="effect__appear sc-bwzfXH iVThND"
  >
    <div
      className="sc-bZQynM hKpJBw"
    >
      <h1
        className="font__accent sc-htpNat ggNPGu"
      >
        Settings
      </h1>
      <div
        className="font__accent sc-EHOje hPfYAa"
      >
        <button
          className="sc-ifAKCX hgRAff"
          data-key="graphics"
          onClick={[Function]}
          tabIndex={0}
        >
          graphics
        </button>
        <button
          className="sc-ifAKCX gXidXF"
          data-key="sound"
          onClick={[Function]}
          tabIndex={0}
        >
          sound
        </button>
        <button
          className="sc-ifAKCX gXidXF"
          data-key="controls"
          onClick={[Function]}
          tabIndex={0}
        >
          controls
        </button>
      </div>
      <div
        className="sc-gzVnrw jyBrwk"
      >
        <button
          className="effect__glitch sc-bxivhb eGYeuD"
          data-text="Back"
          onClick={[Function]}
          tabIndex={-1}
        >
          Back
          <a
            href="/menu"
            onClick={[Function]}
            tabIndex="0"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Settings /> Should renders 1`] = `
<div
  role="button"
  tabIndex="0"
>
  <div
    className="effect__appear sc-bwzfXH iVThND"
  >
    <div
      className="sc-bZQynM hKpJBw"
    >
      <h1
        className="font__accent sc-htpNat ggNPGu"
      >
        Settings
      </h1>
      <div
        className="font__accent sc-EHOje hPfYAa"
      >
        <button
          className="sc-ifAKCX hgRAff"
          data-key="graphics"
          onClick={[Function]}
          tabIndex={0}
        >
          graphics
        </button>
        <button
          className="sc-ifAKCX gXidXF"
          data-key="sound"
          onClick={[Function]}
          tabIndex={0}
        >
          sound
        </button>
        <button
          className="sc-ifAKCX gXidXF"
          data-key="controls"
          onClick={[Function]}
          tabIndex={0}
        >
          controls
        </button>
      </div>
      <div
        className="sc-gzVnrw jyBrwk"
      >
        <button
          className="effect__glitch sc-bxivhb eGYeuD"
          data-text="Back"
          onClick={[Function]}
          tabIndex={-1}
        >
          Back
          <a
            href="/menu"
            onClick={[Function]}
            tabIndex="0"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Settings /> Should reset settings 1`] = `
<div
  role="button"
  tabIndex="0"
>
  <div
    className="effect__appear sc-bwzfXH iVThND"
  >
    <div
      className="sc-bZQynM hKpJBw"
    >
      <h1
        className="font__accent sc-htpNat ggNPGu"
      >
        Settings
      </h1>
      <div
        className="font__accent sc-EHOje hPfYAa"
      >
        <button
          className="sc-ifAKCX hgRAff"
          data-key="graphics"
          onClick={[Function]}
          tabIndex={0}
        >
          graphics
        </button>
        <button
          className="sc-ifAKCX gXidXF"
          data-key="sound"
          onClick={[Function]}
          tabIndex={0}
        >
          sound
        </button>
        <button
          className="sc-ifAKCX gXidXF"
          data-key="controls"
          onClick={[Function]}
          tabIndex={0}
        >
          controls
        </button>
      </div>
      <div
        className="sc-gzVnrw jyBrwk"
      >
        <button
          className="effect__glitch sc-bxivhb eGYeuD"
          data-text="Apply"
          onClick={[Function]}
          tabIndex={-1}
        >
          Apply
          <a
            href="/menu"
            onClick={[Function]}
            tabIndex="0"
          />
        </button>
        <button
          className="effect__glitch sc-bxivhb eGYeuD"
          data-text="Reset"
          onClick={[Function]}
          tabIndex={-1}
        >
          Reset
          <a
            href="/menu"
            onClick={[Function]}
            tabIndex="0"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Settings /> Should reset settings 2`] = `
<div
  role="button"
  tabIndex="0"
>
  <div
    className="effect__appear sc-bwzfXH iVThND"
  >
    <div
      className="sc-bZQynM hKpJBw"
    >
      <h1
        className="font__accent sc-htpNat ggNPGu"
      >
        Settings
      </h1>
      <div
        className="font__accent sc-EHOje hPfYAa"
      >
        <button
          className="sc-ifAKCX hgRAff"
          data-key="graphics"
          onClick={[Function]}
          tabIndex={0}
        >
          graphics
        </button>
        <button
          className="sc-ifAKCX gXidXF"
          data-key="sound"
          onClick={[Function]}
          tabIndex={0}
        >
          sound
        </button>
        <button
          className="sc-ifAKCX gXidXF"
          data-key="controls"
          onClick={[Function]}
          tabIndex={0}
        >
          controls
        </button>
      </div>
      <div
        className="sc-gzVnrw jyBrwk"
      >
        <button
          className="effect__glitch sc-bxivhb eGYeuD"
          data-text="Back"
          onClick={[Function]}
          tabIndex={-1}
        >
          Back
          <a
            href="/menu"
            onClick={[Function]}
            tabIndex="0"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Settings /> Should switch group 1`] = `
<div
  role="button"
  tabIndex="0"
>
  <div
    className="effect__appear sc-bwzfXH iVThND"
  >
    <div
      className="sc-bZQynM hKpJBw"
    >
      <h1
        className="font__accent sc-htpNat ggNPGu"
      >
        Settings
      </h1>
      <div
        className="font__accent sc-EHOje hPfYAa"
      >
        <button
          className="sc-ifAKCX gXidXF"
          data-key="graphics"
          onClick={[Function]}
          tabIndex={0}
        >
          graphics
        </button>
        <button
          className="sc-ifAKCX hgRAff"
          data-key="sound"
          onClick={[Function]}
          tabIndex={0}
        >
          sound
        </button>
        <button
          className="sc-ifAKCX gXidXF"
          data-key="controls"
          onClick={[Function]}
          tabIndex={0}
        >
          controls
        </button>
      </div>
      <div
        className="sc-gzVnrw jyBrwk"
      >
        <button
          className="effect__glitch sc-bxivhb eGYeuD"
          data-text="Back"
          onClick={[Function]}
          tabIndex={-1}
        >
          Back
          <a
            href="/menu"
            onClick={[Function]}
            tabIndex="0"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Settings /> Should update settings and switch to "changed" state 1`] = `
<div
  role="button"
  tabIndex="0"
>
  <div
    className="effect__appear sc-bwzfXH iVThND"
  >
    <div
      className="sc-bZQynM hKpJBw"
    >
      <h1
        className="font__accent sc-htpNat ggNPGu"
      >
        Settings
      </h1>
      <div
        className="font__accent sc-EHOje hPfYAa"
      >
        <button
          className="sc-ifAKCX hgRAff"
          data-key="graphics"
          onClick={[Function]}
          tabIndex={0}
        >
          graphics
        </button>
        <button
          className="sc-ifAKCX gXidXF"
          data-key="sound"
          onClick={[Function]}
          tabIndex={0}
        >
          sound
        </button>
        <button
          className="sc-ifAKCX gXidXF"
          data-key="controls"
          onClick={[Function]}
          tabIndex={0}
        >
          controls
        </button>
      </div>
      <div
        className="sc-gzVnrw jyBrwk"
      >
        <button
          className="effect__glitch sc-bxivhb eGYeuD"
          data-text="Apply"
          onClick={[Function]}
          tabIndex={-1}
        >
          Apply
          <a
            href="/menu"
            onClick={[Function]}
            tabIndex="0"
          />
        </button>
        <button
          className="effect__glitch sc-bxivhb eGYeuD"
          data-text="Reset"
          onClick={[Function]}
          tabIndex={-1}
        >
          Reset
          <a
            href="/menu"
            onClick={[Function]}
            tabIndex="0"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
